server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -511,12 +518,18 @@ private void setRefreshRepositoryUuidOnRestore(boolean refreshRepositoryUuidOnRe
      * Best-effort attempt to make sure that we know all the repository UUIDs. Calls {@link Repository#getRepositoryData} on every
      * {@link BlobStoreRepository} with a missing UUID.
      *
-     * @param enabled If {@code false} this method completes the listener immediately
+     * @param enabled             If {@code false} this method completes the listener immediately
      * @param repositoriesService Supplies the repositories to check
-     * @param onCompletion Action that is executed when all repositories have been refreshed.
+     * @param onCompletion        Action that is executed when all repositories have been refreshed.
+     * @param responseExecutor    Executor on which to execute {@code onCompletion} if not using the calling thread.
```

Comment:
From what I can tell, `responseExecutor` is just used to log L549 and L554, per `getRepositoryData`'s documentation that the executor is used to run the given action listener. `onCompletion` is going to run on the same thread as the method caller. Am I missing something?

Reply:
Yeah the executor is also (sometimes) used to release the ref acquired by the `refs.acquire()` on L556, and `onCompletion` runs when the last such ref is released.

Reply:
Okay, I think I discovered that this works. [AbstractRefCounted will run the runnable from decRef](https://github.com/elastic/elasticsearch/blob/0aa4758f02f736176a0f1c211ca701465ad05f63/libs/core/src/main/java/org/elasticsearch/core/AbstractRefCounted.java#L70). RefCountingRunnable also calls [decRef() on close()](https://github.com/elastic/elasticsearch/blob/75b0c5d13b93cc2092896e75c52af20e5bed3928/server/src/main/java/org/elasticsearch/action/support/RefCountingRunnable.java#L112). So it looks like one of the acquire() ops might run the Runnable, or the close(). At least that's what I gather without digging deeper.

Reply:
Exactly. I mean it couldn't be any other way - we've got to run the `closeInternal()` on _some_ thread, and we don't have any other executors in play, so it has to be the thread that calls `decRef()` for the last time.

Reply:
Sure. It's more of a conceptual relationship than a literal one (not passing an executor to the runnable), at this level of the code.

Reply:
Yeah the point there is that `repository.getRepositoryData()` will mostly complete its listener on the calling thread, which is the right thread to use for `onCompletion` anyway, so we can save the overhead of dispatching that work back to the same executor by passing in the two arguments separately.